crates/java_string/src/char.rs

```diff
@@ -0,0 +1,1012 @@
+use std::char::ParseCharError;
+use std::cmp::Ordering;
+use std::fmt;
+use std::fmt::{Debug, Display, Formatter, Write};
+use std::hash::{Hash, Hasher};
+use std::iter::{once, FusedIterator, Once};
+use std::ops::Range;
+use std::str::FromStr;
+
+use crate::validations::{TAG_CONT, TAG_FOUR_B, TAG_THREE_B, TAG_TWO_B};
+
+// JavaCodePoint is guaranteed to have the same repr as a u32, with valid values
+// of between 0 and 0x10FFFF, the same as a unicode code point. Surrogate code
+// points are valid values of this type.
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[repr(C)]
+pub struct JavaCodePoint {
```

Comment:
I named it as such to avoid confusion with Java's `char` type, which only supports values between 0 and 0xffff, whereas "code point" means the same thing in both Java and Rust.